nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,378 @@
+          };
+        };
+      }
+
+      {
+        name = "bupstash_posthook-${name}";
+        value = {
+          description = "Hook to run after the ${name} bupstash backup.";
+
+          partOf = [ "bupstash-${name}.target" ];
+          requiredBy = [ "bupstash-${name}.target" ];
+
+          # Require the preHook Unit, so that in case of failure it does not try to run
+          requires = [ "bupstash_prehook-${name}.service" ];
+
+          script = ''
+            echo "Running ${name} bupstash postHook"
+
+            ${job.postHook}
+          '';
+
+          serviceConfig = {
+            Type = "oneshot";
+            RuntimeDirectory = "bupstash";
+            StateDirectory = "bupstash";
+            WorkingDirectory = "/var/lib/bupstash";
+          };
+        };
+      }
+    ]
+    ++
+      builtins.map
+        (
+          r:
+          let
+            repo = builtins.replaceStrings [ "/" ] [ "_" ] (
+              if builtins.isString r then r else "ssh-${r.server}-${r.path}"
+            );
+          in
+          nameValuePair "bupstash-${name}-${repo}" {
+            path = [ pkgs.coreutils ] ++ (optional (builtins.isAttrs r) pkgs.openssh);
+
+            after = [ "bupstash_prehook-${name}.service" ];
+            before = [ "bupstash_posthook-${name}.service" ];
+
+            partOf = [ "bupstash-${name}.target" ];
+            wantedBy = [ "bupstash-${name}.target" ];
+
+            preStart = ''
+              # We assume that a repository is already initialized iff its log exists
+              if [ ! -f "sendlog" ]; then
+                ${bupstash} init
+              fi
+            '';
+
+            script = ''
+              ${bupstash} put ${optionalString (job.command != null) "--exec"} \
+                ${builtins.toString job.extraOptions} \
+                ${builtins.toString (mapAttrsToList (name: value: "${name}='${value}'") job.tags)} \
+                :: ${if (job.command == null) then builtins.toString job.paths else job.command}
+            '';
+
+            environment =
+              {
+                BUPSTASH_KEY = job.key;
+                BUPSTASH_SEND_LOG = "/var/lib/bupstash/${name}.${repo}/sendlog";
+              }
+              // (
+                if (builtins.isAttrs r) then
+                  {
+                    BUPSTASH_REPOSITORY_COMMAND = "ssh -i ${r.privateKey} ${r.user}@${r.server} bupstash serve ${r.path}";
+                  }
+                else
+                  { BUPSTASH_REPOSITORY = r; }
+              );
+
+            serviceConfig = {
+              Type = "oneshot";
+              RuntimeDirectory = "bupstash/${name}.${repo}";
+              StateDirectory = "bupstash/${name}.${repo}";
+              WorkingDirectory = "/var/lib/bupstash/${name}.${repo}";
+              DynamicUser = true;
+              User = mkIf (job.user != null) job.user;
+              Group = mkIf (job.group != null) job.group;
```

Comment:
Can't this create problems when no `bupstash` user really exists and then we have different users with the same same and different uids

Reply:
With a dynamic user it will only set the pretty name. In case a static user with that name exists, DynamicUser= wil be disabled automatically. At least that is how I understand it.